src/CSE.cpp

```diff
@@ -61,9 +61,12 @@ bool should_extract(const Expr &e) {
     }
 
     if (const Call *a = e.as<Call>()) {
-        if (!a->is_pure()) {
-            //non-pure Calls may have side-effects, thus
-            //may not be re-ordered or reduced in number.
+        if (!a->is_pure() && (a->call_type != Call::Halide)) {
```

Comment:
It seems Halide calls are conceptually "pure" if we model any modifications of buffers as an output of the function. (Which I think is why the CSE is allowed here.) Is it possible to move this case into is_pure? If not, perhaps we should add another method alongside is_pure to handle this case.